### PR TITLE
Function to return Total Energy in physics/mechanics

### LINKED ISSUE #16201: Function to return Total Energy in physics/mechanics
At present, there is are functions to calculate kinetic energy and potential energy of a rigid body in ```sympy/physics/mechanics/rigidbody.py``` but no function to return the total energy of the rigid body. I propose the following PR. Please let me know your views about it.

### PR DESCRIPTION
**Pull request by ritesh99rakesh**
Fixes #16201 
Closes #16202

#### Brief description of what is fixed or changed
At present, we have function to get the kinetic energy and potential energy of rigid body, but we do not have any function to give the total energy of the rigid body. This PR adds a function under ```sympy/physics/mechanics/rigidbody.py``` to return the total energy of the body. 
I have defined total_energy = kinetic_energy + potential_energy

#### Other comments
Please review it and comment your suggestions.

#### Release Notes

* physics.mechanics
    * added functions total_energy, rotational_kinetic_energy, and translational_kinetic_energy, in rigidbody.py
    * added function total_energy in particle.py

### CODE REVIEW

File: sympy/physics/mechanics/tests/test_particle.py [hunk 1]
```diff
@@ -38,6 +38,8 @@ def test_particle():
     assert p.angular_momentum(O, N) == m2 * r * (v3 * N.x - v1 * N.z)
     p.potential_energy = m * g * h
     assert p.potential_energy == m * g * h
+    assert p.total_energy(
+        N) == m*g*h + m2 * v1**2 / 2 + m2 * v2**2 / 2 + m2 * v3**2 / 2
     # TODO make the result not be system-dependent
     assert p.kinetic_energy(
         N) in [m2*(v1**2 + v2**2 + v3**2)/2,
```
Comment (sudoWin) [inline sympy/physics/mechanics/tests/test_particle.py]: You can add more spacing between operators.

File: sympy/physics/mechanics/particle.py [hunk 2]
```diff
@@ -227,6 +227,33 @@ def potential_energy(self, scalar):
 
         self._pe = sympify(scalar)
 
+    def total_energy(self, frame):
+        """The total energy of the Particle.
+        The total energy, E, of a particle, P, is given by
+        'E = K.E + P.E.'
+        where P.E. is the potential energy of the particle P, and
+        K.E. is kinetic energy of the particle P, in the supplied ReferenceFrame.
```
Comment (czgdp1807) [inline sympy/physics/mechanics/particle.py:235]:
```suggestion
        K.E. is kinetic energy of the particle P, in the supplied ReferenceFrame.

```

File: sympy/physics/mechanics/particle.py [hunk 3]
```diff
@@ -227,6 +227,33 @@ def potential_energy(self, scalar):
...
+        Parameters
+        ==========
```
Comment (czgdp1807) [inline sympy/physics/mechanics/particle.py:237]:
```suggestion
        ==========

```

File: sympy/physics/mechanics/particle.py [hunk 4]
```diff
@@ -227,6 +227,33 @@ def potential_energy(self, scalar):
...
+        frame : ReferenceFrame
+            The Particle's velocity is typically defined with respect to
+            an inertial frame but any relevant frame in which the velocity is
+            known can be supplied.
```
Comment (czgdp1807) [inline sympy/physics/mechanics/particle.py:241]:
```suggestion
            known can be supplied.
 
```

File: sympy/physics/mechanics/particle.py [hunk 5]
```diff
@@ -227,6 +227,33 @@ def potential_energy(self, scalar):
...
+        Examples
+        ========
```
Comment (czgdp1807) [inline sympy/physics/mechanics/particle.py:243]:
```suggestion
        ========

```

File: sympy/physics/mechanics/particle.py [hunk 6]
```diff
@@ -227,6 +227,33 @@ def potential_energy(self, scalar):
 
         self._pe = sympify(scalar)
 
+    def total_energy(self, frame):
+        """The total energy of the Particle.
```
Comment (czgdp1807) [inline sympy/physics/mechanics/particle.py:231]:
```suggestion
        """
        The total energy of the Particle.
```

File: sympy/physics/mechanics/rigidbody.py [hunk 7]
```diff
@@ -202,6 +202,72 @@ def angular_momentum(self, point, frame):
...
+        >>> b.set_ang_vel(N, omega * b.x)
+        >>> P = Point('P')
+        >>> P.set_vel(N, v * N.x)
+        >>> I = outer (b.x, b.x)
+        >>> inertia_tuple = (I, P)
+        >>> B = RigidBody('B', P, b, M, inertia_tuple)
+        >>> B.rotational_kinetic_energy(N)
+        omega**2/2
+        """
+
```
Comment (czgdp1807) [inline sympy/physics/mechanics/rigidbody.py:235]:
```suggestion
    def rotational_kinetic_energy(self, frame):
        """
        Rotational kinetic energy of the rigid body
        The rotational kinetic energy, T, of a rigid body, B, is given by
        'T = 1/2 (I omega^2)'
        where I is the central inertia dyadic of rigid body B,
        omega is the angular velocity of the body's
        mass center in the supplied ReferenceFrame.

        Parameters
        =========

        frame : ReferenceFrame
            The RigidBody's angular velocity and the velocity of it's mass
            center are typically defined with respect to an inertial frame but
            any relevant frame in which the velocities are known can be supplied.

        Examples
        ========

        >>> from sympy.physics.mechanics import Point, ReferenceFrame, outer
        >>> from sympy.physics.mechanics import RigidBody
        >>> from sympy import symbols
        >>> M, v, r, omega = symbols('M v r omega')
        >>> N = ReferenceFrame('N')
        >>> b = ReferenceFrame('b')
        >>> b.set_ang_vel(N, omega * b.x)
        >>> P = Point('P')
        >>> P.set_vel(N, v * N.x)
        >>> I = outer (b.x, b.x)
        >>> inertia_tuple = (I, P)
        >>> B = RigidBody('B', P, b, M, inertia_tuple)
        >>> B.rotational_kinetic_energy(N)
        omega**2/2
        """
```

File: sympy/physics/mechanics/rigidbody.py [hunk 8]
```diff
@@ -202,6 +202,72 @@ def angular_momentum(self, point, frame):
...
+        >>> b.set_ang_vel(N, omega * b.x)
+        >>> P = Point('P')
+        >>> P.set_vel(N, v * N.x)
+        >>> I = outer (b.x, b.x)
+        >>> inertia_tuple = (I, P)
+        >>> B = RigidBody('B', P, b, M, inertia_tuple)
+        >>> B.translational_kinetic_energy(N)
+        M*v**2/2
+        """
+
```
Comment (czgdp1807) [inline sympy/physics/mechanics/rigidbody.py:268]:
```suggestion
        """
        Translational kinetic energy of the rigid body
        The translational kinetic energy, T, of a rigid body, B, is given by
        'T = 1/2 (m v^2)'
        where m is the mass of rigid body B, v is the velocity of the body's
        mass center in the supplied ReferenceFrame.

        Parameters
        =========

        frame : ReferenceFrame
            The RigidBody's angular velocity and the velocity of it's mass
            center are typically defined with respect to an inertial frame but
            any relevant frame in which the velocities are known can be supplied.

        Examples
        ========

        >>> from sympy.physics.mechanics import Point, ReferenceFrame, outer
        >>> from sympy.physics.mechanics import RigidBody
        >>> from sympy import symbols
        >>> M, v, r, omega = symbols('M v r omega')
        >>> N = ReferenceFrame('N')
        >>> b = ReferenceFrame('b')
        >>> b.set_ang_vel(N, omega * b.x)
        >>> P = Point('P')
        >>> P.set_vel(N, v * N.x)
        >>> I = outer (b.x, b.x)
        >>> inertia_tuple = (I, P)
        >>> B = RigidBody('B', P, b, M, inertia_tuple)
        >>> B.translational_kinetic_energy(N)
        M*v**2/2
        """
```

File: sympy/physics/mechanics/rigidbody.py [hunk 9]
```diff
@@ -300,6 +360,37 @@ def potential_energy(self, scalar):
 
         self._pe = sympify(scalar)
 
+    def total_energy(self, frame):
+        """The total energy of the RigidBody.
```
Comment (czgdp1807) [inline sympy/physics/mechanics/rigidbody.py:364]:
```suggestion
        """
        The total energy of the RigidBody.
```

File: sympy/physics/mechanics/rigidbody.py [hunk 10]
```diff
@@ -300,6 +360,37 @@ def potential_energy(self, scalar):
...
+        The total energy E, of rigidbody B, is given by
+        E = K.E. + P.E.
+        where P.E. is the potential energy of the rigid body B and
+        K.E. is kinetic energy of the rigid body B in the supplied ReferenceFrame.
```
Comment (czgdp1807) [inline sympy/physics/mechanics/rigidbody.py:368]:
```suggestion
        K.E. is kinetic energy of the rigid body B in the supplied ReferenceFrame.

```

File: sympy/physics/mechanics/rigidbody.py [hunk 11]
```diff
@@ -300,6 +360,37 @@ def potential_energy(self, scalar):
...
+        Parameters
+        ==========
```
Comment (czgdp1807) [inline sympy/physics/mechanics/rigidbody.py:370]:
```suggestion
        ==========

```

File: sympy/physics/mechanics/rigidbody.py [hunk 12]
```diff
@@ -300,6 +360,37 @@ def potential_energy(self, scalar):
...
+        frame : ReferenceFrame
+            The RigidBody's angular velocity and the velocity of it's mass
+            center are typically defined with respect to an inertial frame but
+            any relevant frame in which the velocities are known can be supplied.
+        Examples
+        ========
```
Comment (czgdp1807) [inline sympy/physics/mechanics/rigidbody.py:376]:
```suggestion
            any relevant frame in which the velocities are known can be supplied.

        Examples
        ========

```

**Comment by oscarbenjamin** [comment]
There is now another PR that duplicates some of the work here https://github.com/sympy/sympy/pull/20016

**Comment by moorepants** [comment]
My opinion is that this is not needed, but if it is added it doesn't hurt anything.

**Comment by oscarbenjamin** [comment]
We need to make a decision about whether we want this or not. If it isn't needed then we should close both PRs.

Personally I find the API a bit weird:
```
        >>> B = RigidBody('B', P, b, M, Inertia_tuple)
        >>> B.potential_energy = M * g * h
        >>> B.total_energy(N)
```
What's the point in assigning an expression to the potential energy just so it can be added to the kinetic energy? You could just do:
```
    >>> potential = M * g * h
    >>> B.kinetic_energy(N) + potential 
```
Also should the potential energy not in general be a function of coordinates from some frame?

**Comment by moorepants** [comment]
> We need to make a decision about whether we want this or not. If it isn't needed then we should close both PRs.

I would just say no, let's close them.

> What's the point in assigning an expression to the potential energy just so it can be added to the kinetic energy?

This is because the LagrangesMethod needs to be aware of the potential energy of any particle or rigidbody, so the api was designed that you attach the expression those objects. It is just what it is.

**Comment by oscarbenjamin** [comment]
Okay, I'll close this. @ritesh99rakesh thank you for working on this but I don't think it is needed.

**Pull request closed without merging**